stelaval/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,176 @@
+package assignment2;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class BinaryTree<T> {
+
+  private Node<T> root;
+  private Set<Node<T>> nodeSet;
+
+  public BinaryTree() {
+    nodeSet = new HashSet<>();
+  }
+
+  /**
+   * The nodes of the binary tree are stored in a set in order to avoid duplicate values.
+   * 
+   * @param newNode is the node to be added to the set of nodes
+   */
+  public void add(Node<T> newNode) {
+    if (root == null) {
+      root = newNode;
+    }
+
+    if (nodeSet.contains(newNode)) {
+      throw new IllegalArgumentException();
+    } else {
+      nodeSet.add(newNode);
+    }
+  }
+
+  /**
+   * Prints all the ancestors of a node
+   * 
+   * @param key is a node which ancestors are printed
+   * 
+   * @throws Exception if given key is not present in a tree
+   */
+  public void printAncestors(Node<T> key) throws Exception {
+    if (nodeSet.contains(key)) {
+      containsKey(root, key);
+    } else {
+      throw new Exception("Key is out of tree");
+    }
+  }
+
+  /**
+   * Checks if a node is present in a subtree

```

Comment:
The description (and method name) is not exactly correct - checking if node is present in subtree is just part of what it does. More importantly it prints things out.

It would be a bit more clear if method name was `printAncestorsIfNodeInSubtree` (or `printAncestors`, or something similar) and comment contained more detailed description and the `@returns` line that would explain what is the return value.